src/uct/ib/base/ib_iface.c

```diff
@@ -740,17 +742,19 @@ static ucs_status_t uct_ib_iface_search_pkey(uct_ib_iface_t *iface,
             continue;
         }
 
-        if ((config->pkey == UCS_HEXUNITS_AUTO) &&
-            !(pkey & UCT_IB_PKEY_MEMBERSHIP_MASK)) {
-            ucs_trace("skipping send-only pkey[%d]=0x%x on "UCT_IB_IFACE_FMT,
-                      pkey_index, pkey, UCT_IB_IFACE_ARG(iface));
-            has_pkey_limited = 1;
-            continue;
-        }
-
         /* take only the lower 15 bits for the comparison */
         if ((config->pkey == UCS_HEXUNITS_AUTO) ||
             ((pkey & UCT_IB_PKEY_PARTITION_MASK) == config->pkey)) {
+            if (!(pkey & UCT_IB_PKEY_MEMBERSHIP_MASK) &&
+                (/* limited PKEY hasn't been find yet */
+                 (lim_pkey == UCT_IB_ADDRESS_INVALID_PKEY) ||
+                 /* find more suitable PKEY with limited membership */
+                 ((pkey & UCT_IB_PKEY_PARTITION_MASK) == config->pkey))) {
```

Comment:
can we unit this if with the if in lines 746-747? seems there is some duplication

Reply:
reworked to make it simpler
could you check pls?